Domains: Fix site-switching issue in domain management

drewblaisdell opened the pull request:
`SitesList` emits a change event when navigating to a route that isn't for the given site, or when navigating from a route that selects a site to an all-sites route. This means that the given `selectedSite` property is false for a render when going from a route that expects a site to be selected to an all-sites route. Many of the domain components currently assume `selectedSite` (and the data associated with the currently selected site) is present, and throw an error when navigating to an all-sites page (e.g. the Reader).

This PR removes the `data-observe` mixin from the following data components to ensure that a false `selectedSite` isn't passed to their children when switching sites:

- `DomainManagementData`
- `EmailData`
- `TransferData`
- `WhoisData`

We are sure that `sites` is loaded and that a site is selected when a domains component mounts in these cases, and site-switching is accompanied by a route change which triggers a re-render, so we don't need `data-observe` here.

**Testing**
- Visit the following routes:
   - `/domains/manage/:site/edit/:domain`
   - `/domains/manage/:site/email/:domain`
   - `/domains/manage/:site/add-google-apps/:domain`
   - `/domains/manage/:site/contacts-privacy/:domain`
   - `/domains/manage/:site/edit-contact-info/:domain`
   - `/domains/manage/:site/privacy-protection/:domain`
  - `/domains/manage/:site/transfer/:domain`
- Click 'Reader' in the masterbar.
- Assert that no error appears in the console and that you are taken to the Reader.

- [x] QA review
- [x] Code review

Props to @scruffian for pointing this out in #579.

scruffian commented:
> are there cases when sites would change and the route wouldn't change that this will miss?

The one that springs to mind is empty localStorage. If you clear local storage and reload http://calypso.localhost:3000/domains/manage/:site/edit/:domain, then you are redirect to an all sites picker, which is a regression.

drewblaisdell commented:
Ah, thanks for pointing that out @scruffian.

It looks like these components need to expect a missing `selectedSite` prop, because `SitesList` updates whenever the selected site changes, triggering a render in `StoreConnection` when `selectedSite` becomes false before the next component is mounted. I updated the commit to guard against a false `selectedSite` in the affected domain management components, and updated the first comment with the reason for this change and new testing instructions.

gziolo commented:
I just noticed similar issue when cancelling purchase (domain or domain mapping):
> Warning: Failed propType: Invalid prop `selectedSite` of type `boolean` supplied to `ConfirmCancelPurchaseLoadingPlaceholder`, expected `object`. Check the render method of `StoreConnection`.

scruffian commented:
>  If you clear local storage and reload http://calypso.localhost:3000/domains/manage/:site/edit/:domain, then you are redirect to an all sites picker, which is a regression.

This still happens on this PR

scruffian commented:
I added the original commit back in 7e76ff5, since this is simpler and reduces our reliance on `data-observe`.

drewblaisdell commented:
I updated the commit to not observe `SitesList` in the data components that wrap components that don't handle this properly, and updated the testing instructions.

scruffian commented:
LGTM :ship:
